File: plugins/lofi/stutter/stutter.js
// Stutter — beat-repeat. Captures the first slice of every cycle into a
// buffer, then plays that slice back N-1 times before re-arming. Loops
// forever as the cycle wraps, so a held vowel turns into a percussive
// chop. Mix lets you blend live signal under the repeats; at mix=1 the
// stutter fully replaces the dry path.

function StutterProcessor() {
  // 1.5 s max slice = 72 000 samples @ 48 kHz, round up for headroom.
  this.buf = new Array(96000);
  for (var k = 0; k < this.buf.length; k++) this.buf[k] = 0;
  this.cursor = 0;
}
StutterProcessor.parameterDescriptors = [
  { name: 'slice',  label: 'Slice (ms)', defaultValue: 120, minValue: 20, maxValue: 500 },
  { name: 'repeat', label: 'Repeats',    defaultValue: 4,   minValue: 1,  maxValue: 8 },
  { name: 'mix',    label: 'Mix',        defaultValue: 0.7, minValue: 0,  maxValue: 1 }
];
StutterProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var sliceMs = parameters.slice[0];
  var repeats = Math.max(1, Math.floor(parameters.repeat[0]));
  var mix = parameters.mix[0];
  var sliceLen = Math.max(1, Math.floor(sampleRate * sliceMs / 1000));
  if (sliceLen > this.buf.length) sliceLen = this.buf.length;
  var totalLen = sliceLen * repeats;
  for (var i = 0; i < input.length; i++) {
    var phase = this.cursor % totalLen;
    if (phase < sliceLen) {
      // First pass through the cycle: record and pass dry.
      this.buf[phase] = input[i];
      output[i] = input[i];
    } else {
      var inSlice = phase % sliceLen;
      var wet = this.buf[inSlice];
      output[i] = input[i] * (1 - mix) + wet * mix;
    }
    this.cursor++;
  }
  return true;
};

registerProcessor('stutter', StutterProcessor);

File: plugins/lofi/stutter/Stutter.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Stutter (native port) — beat-repeat. First slice of each cycle is
// captured into a buffer and replayed N-1 times before re-arming.
// Held vowels turn into percussive chops, drums into glitch fills.
public final class Stutter implements VocalMonitorNativePlugin {
    private float[] buf;
    private int cursor = 0;
    private int sampleRate = 44100;
    private float slice = 120f, repeat = 4f, mix = 0.7f;

    @Override
    public void init(int sr) {
        this.sampleRate = sr;
        buf = new float[96000];
        cursor = 0;
    }

    @Override public String[] parameterNames() { return new String[] { "slice", "repeat", "mix" }; }
    @Override public float parameterMin(String n) {
        switch (n) {
            case "slice":  return 20f;
            case "repeat": return 1f;
            default:       return 0f;
        }
    }
    @Override public float parameterMax(String n) {
        switch (n) {
            case "slice":  return 500f;
            case "repeat": return 8f;
            default:       return 1f;
        }
    }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "slice":  return 120f;
            case "repeat": return 4f;
            case "mix":    return 0.7f;
            default:       return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "slice":  return "Slice (ms)";
            case "repeat": return "Repeats";
            case "mix":    return "Mix";
            default:       return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "slice":  slice = v; break;
            case "repeat": repeat = v; break;
            case "mix":    mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        int sliceLen = Math.max(1, (int) Math.floor(sampleRate * slice / 1000f));
        if (sliceLen > buf.length) sliceLen = buf.length;
        final int repeats = Math.max(1, (int) Math.floor(repeat));
        final int totalLen = sliceLen * repeats;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final float[] b = buf;
        int c = cursor;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            int phase = c % totalLen;
            if (phase < sliceLen) {
                b[phase] = input[i];
                output[i] = input[i];
            } else {
                int inSlice = phase % sliceLen;
                float wet = b[inSlice];
                output[i] = input[i] * dry + wet * mixLocal;
            }
            c++;
        }
        cursor = c;
    }
}
